### PR TITLE
fix for IPP 2021.10 with OneAPI 2024

### LINKED ISSUE #25270: support for IPP 2021.10 in OneAPI 2024
### Describe the feature and motivation

IPP 2021.10 change the directory structure, need some tune for find the IPP in oneapi 

**OpenCVFindIPP.cmake**

```
-  set(IPP_INCLUDE_DIRS ${IPP_ROOT_DIR}/include)
+  set(IPP_INCLUDE_DIRS ${IPP_ROOT_DIR}/include/ipp)

   set(__msg)
   if(EXISTS ${IPP_ROOT_DIR}/include/ippicv_redefs.h)
@@ -118,7 +118,7 @@ macro(ipp_detect_version)
   if(APPLE AND NOT HAVE_IPP_ICV)
     _ipp_set_library_dir(${IPP_ROOT_DIR}/lib)
   elseif(IPP_X64)
-    _ipp_set_library_dir(${IPP_ROOT_DIR}/lib/intel64)
+    _ipp_set_library_dir(${IPP_ROOT_DIR}/lib/)
   else()
     _ipp_set_library_dir(${IPP_ROOT_DIR}/lib/ia32)
   endif()
@@ -271,7 +271,7 @@ if(NOT DEFINED IPPROOT)
 endif()

 file(TO_CMAKE_PATH "${IPPROOT}" __IPPROOT)
-if(EXISTS "${__IPPROOT}/include/ippversion.h")
+if(EXISTS "${__IPPROOT}/include/ipp/ippversion.h")
   set(IPP_ROOT_DIR ${__IPPROOT})
   ipp_detect_version()
 endif()

```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by junxnone**
fixes #25270

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
```
-- IPPICV: Downloading ippicv_2021.10.1_lnx_intel64_20231206_general.tgz from https://raw.githubusercontent.com/opencv/opencv_3rdparty/c7c6d527dde5fee7cb914ee9e4e20f7436aab3a1/ippicv/ippicv_2021.10.1_lnx_intel64_20231206_general.tgz
CMake Error at cmake/OpenCVFindIPP.cmake:87 (cmake_path):
  Unknown CMake command "cmake_path".
Call Stack (most recent call first):
  cmake/OpenCVFindIPP.cmake:278 (ipp_detect_version)
  cmake/OpenCVFindLibsPerf.cmake:12 (include)
  CMakeLists.txt:820 (include)


-- Configuring incomplete, errors occurred!
```

**Comment by junxnone** [comment]
Sorry,  `cmake_path()` is new in version 3.20. 

**But opencv is use the min 3.5.1:**
https://github.com/opencv/opencv/blob/13c95efa74690cace66316f04894ce63b9d11570/cmake/OpenCVMinDepVersions.cmake#L2 

Will try to get new method.

**Comment by asmorkalov** [comment]
Looks like the same changes are required for IPP IW in cmake/OpenCVFindIPPIW.cmake.

**Comment by junxnone** [comment]
> Looks like the same changes are required for IPP IW in cmake/OpenCVFindIPPIW.cmake.

not found

- The file `ippicv_redefs.h` is not release with `IPP(OneAPI)`. So no matter which IPP is used, `IPP within OneAPI` or  `ICV(IPP for OpenCV)`, they all take the IW from ICV package.

https://github.com/opencv/opencv/blob/b0d905864d6af8fef33ab30e102aa3cfcb329fe0/cmake/OpenCVFindIPP.cmake#L90-L92

https://github.com/opencv/opencv/blob/d0ec39206a20c124f91a508fe95884ffc931fbcc/cmake/OpenCVFindIPPIW.cmake#L172-L174

- Is the `IW` from `IPP(OneAPI)` not works for OpenCV?

**Comment by opencv-alalek** [comment]
AFAIK, OpenCV uses "own" IW wrapper due to API from "iw/iw_ll.h" file which is not available in common IPP releases.
Need to check if this API is available in modern IPP package.

**Comment by asmorkalov** [comment]
@opencv-alalek Does it mean, that the PR may be merged?

File: cmake/OpenCVFindIPP.cmake [hunk 1]
```diff
@@ -84,7 +84,7 @@ endmacro()
 # This macro uses IPP_ROOT_DIR variable
 # TODO Cleanup code after ICV package stabilization
 macro(ipp_detect_version)
-  set(IPP_INCLUDE_DIRS ${IPP_ROOT_DIR}/include)
+  get_filename_component(IPP_INCLUDE_DIRS ${IPP_VERSION_FILE} PATH)
```
Comment (opencv-alalek) [inline cmake/OpenCVFindIPP.cmake:87]: Check on the line 93 may fail and fail IPP detection completelly.

Comment (junxnone) [reply cmake/OpenCVFindIPP.cmake:87]: https://github.com/opencv/opencv/blob/c577b2440643bf98942cfd8c49478c545c3fa56b/cmake/OpenCVFindIPP.cmake#L90-L97

> In OneAPI 2024, `inlcude/ipp.h` is the only one that has not moved to the directory `include/ipp/`. 

```
$ tree  /opt/intel/oneapi/ipp/2021.11/include/ -L 2
/opt/intel/oneapi/ipp/2021.11/include/
├── ipp
│   ├── ippbase.h
...
│   ├── ippvm.h
│   ├── iw
│   └── iw++
└── ipp.h

```

Comment (opencv-alalek) [reply cmake/OpenCVFindIPP.cmake:87]: Thank you for the information!

Perhaps we should not change `IPP_INCLUDE_DIRS` to the nested directory. Or add both locations.

I see these IPP includes in OpenCV:

```
./modules/core/include/opencv2/core/private.hpp:220:#include "ippversion.h"
./modules/core/include/opencv2/core/private.hpp:230:#include "ippicv.h"
./modules/core/include/opencv2/core/private.hpp:232:#include "ipp.h"
```

/cc @eplankin

Comment (opencv-alalek) [reply cmake/OpenCVFindIPP.cmake:87]: Well, top-level ipp.h is just a shim:

```
#include <ipp/ipp.h>
```

**Review by opencv-alalek — APPROVED** [review]
- [x] Validated with `l_ipp_oneapi_p_2021.11.0.532_offline.sh`

**Pull request merged**